src/openvslam_ros.cc

```diff
@@ -80,6 +84,43 @@ void system::setParams() {
     transform_tolerance_ = node_->declare_parameter("transform_tolerance", transform_tolerance_);
 }
 
+void system::init_pose_callback(
+    const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg) {
+    Eigen::Matrix4d cam_pose_ros(Eigen::Matrix4d::Identity());
+    Eigen::Vector3d pose_v(
+        msg->pose.pose.position.x,
+        msg->pose.pose.position.y,
+        msg->pose.pose.position.z);
+    cam_pose_ros.block<3, 1>(0, 3) = pose_v;
+
+    Eigen::Quaterniond rot_q(
+        msg->pose.pose.orientation.w,
+        msg->pose.pose.orientation.x,
+        msg->pose.pose.orientation.y,
+        msg->pose.pose.orientation.z);
+    Eigen::Matrix3d rot_m = rot_q.toRotationMatrix();
+    cam_pose_ros.block<3, 3>(0, 0) = rot_m;
+
+    Eigen::Matrix4d rot_ros_to_cv_map_frame;
+    rot_ros_to_cv_map_frame << 0, 0, 1, 0,
+        -1, 0, 0, 0,
+        0, -1, 0, 0,
+        0, 0, 0, 1;
+
+    // Transforming initial pose from ROS coordinate system to CV coordinate system as follows:
+    // T(map, pose_cv) = T(map, pose) * T(pose, pose_cv)
+    // T(map, pose_cv) = T(map, map_cv) * T(map_cv, pose_cv)
+    // where T(pose, pose_cv) == T(map, map_cv)
+    // ||
+    // \/
+    // T(map_cv, pose_cv) = T(map, map_cv).inv * T(map, pose) * T(map, map_cv)
+    Eigen::Matrix4d cam_pose_cv = rot_ros_to_cv_map_frame.transpose() * cam_pose_ros * rot_ros_to_cv_map_frame;
+
+    if (!SLAM_.update_pose(cam_pose_cv)) {
+        RCLCPP_ERROR(node_->get_logger(), "Can not set initial pose");
+    }
```

Comment:
I haven't checked it works, but I think the following process is necessary. The frame_id of initialpose is determined by the fixed frame of rviz2.

```c++
    Eigen::Translation3d trans(
        msg->pose.pose.position.x,
        msg->pose.pose.position.y,
        msg->pose.pose.position.z);
    Eigen::Quaterniond rot_q(
        msg->pose.pose.orientation.w,
        msg->pose.pose.orientation.x,
        msg->pose.pose.orientation.y,
        msg->pose.pose.orientation.z);
    Eigen::Affine3d initialpose_affine(trans * rot_q);

    Eigen::Matrix3d rot_cv_to_ros_map_frame;
    rot_cv_to_ros_map_frame << 0, -1, 0,
        0, 0, -1,
        1, 0, 0;

    Eigen::Affine3d base_link_to_camera_affine;
    try {
        auto base_link_to_camera = tf_->lookupTransform(
            base_link_, camera_link_, tf2_ros::fromMsg(msg->header.stamp),
            tf2::durationFromSec(0.0));
        base_link_to_camera_affine = tf2::transformToEigen(base_link_to_camera.transform);
    }
    catch (tf2::TransformException& ex) {
        RCLCPP_ERROR(node_->get_logger(), "Transform failed: %s", ex.what());
        return;
    }

    Eigen::Affine3d map_to_initialpose_parent_frame_affine;
    try {
        auto map_to_initialpose_frame = tf_->lookupTransform(
            map_frame_, msg->header.frame_id, tf2_ros::fromMsg(msg->header.stamp),
            tf2::durationFromSec(0.0));
        map_to_initialpose_parent_frame_affine = tf2::transformToEigen(map_to_initialpose_frame.transform);
    }
    catch (tf2::TransformException& ex) {
        RCLCPP_ERROR(node_->get_logger(), "Transform failed: %s", ex.what());
        return;
    }

    // Target transform is map_cv -> camera_link and known parameters is following:
    //   rot_cv_to_ros_map_frame(map_cv -> map)
    //   map_to_initialpose_frame(map -> `msg->header.frame_id`)
    //   initialpose_affine(`msg->header.frame_id` -> base_link)
    //   base_link_to_camera_affine(base_link -> camera_link)
    // The flow of the transformation is as follows:
    //   map_cv -> map -> `msg->header.frame_id` -> base_link -> camera_link
    Eigen::Matrix4d initialpose_cv = (rot_cv_to_ros_map_frame * map_to_initialpose_parent_frame_affine
                                      * initialpose_affine * base_link_to_camera_affine)
                                         .matrix();

    if (!SLAM_.update_pose(initialpose_cv)) {
        RCLCPP_ERROR(node_->get_logger(), "Can not set initial pose");
    }
```

Reply:
I also thought about it. In the case if `frame_id` of selected message differs from `map` (I did not saw it on practice, but anyway) we do need this. Thank you for the suggestion. I'll add it into next PR update.

Reply:
The suggestions from above were met in PR [update](https://github.com/OpenVSLAM-Community/openvslam_ros/pull/29/commits/027e79aa7783a8c8bc95beb8732ac4980044605e) commit.